packages/@sanity/form-builder/src/inputs/PortableText/toolbar/BlockStyleSelect.tsx

```diff
@@ -0,0 +1,169 @@
+  // Change the border color variable used by BlockQuote
+  // to make the border visible when the MenuItem is selected
+  &[data-selected] {
+    [data-option='blockquote'] {
+      --card-border-color: var(--card-muted-fg-color);
+    }
+  }
+`
+
+const MENU_POPOVER_PROPS: MenuButtonProps['popover'] = {
+  constrainSize: true,
+  placement: 'bottom-start',
+  portal: 'default',
+}
+
+const TEXT_STYLE_OPTIONS = {
+  h1: (title) => <Heading1>{title}</Heading1>,
+  h2: (title) => <Heading2>{title}</Heading2>,
+  h3: (title) => <Heading3>{title}</Heading3>,
+  h4: (title) => <Heading4>{title}</Heading4>,
+  h5: (title) => <Heading5>{title}</Heading5>,
+  h6: (title) => <Heading6>{title}</Heading6>,
+  normal: (title) => <Normal>{title}</Normal>,
+  blockquote: (title) => <BlockQuote data-option="blockquote">{title}</BlockQuote>,
+}
+
+const TEXT_STYLE_KEYS = Object.keys(TEXT_STYLE_OPTIONS)
+
+const preventDefault = (event: React.MouseEvent<HTMLButtonElement>) => event.preventDefault()
+
+const emptyStyle: BlockStyleItem = {
+  key: 'style-none',
+  style: null,
+  styleComponent: null,
+  title: 'No style',
+}
+
+export const BlockStyleSelect = memo(function BlockStyleSelect(
+  props: BlockStyleSelectProps
+): JSX.Element {
+  const {disabled, items: itemsProp} = props
+  const editor = usePortableTextEditor()
+  const features = useFeatures()
+  const focusBlock = useFocusBlock()
+  const [changed, setChanged] = useState(false)
+
+  const _disabled =
+    disabled || (focusBlock ? features.types.block.name !== focusBlock._type : false)
+
+  // @todo: Explain what this does
+  const activeKeys = useActiveStyleKeys({items: itemsProp})
+
+  const {activeItems, items} = useMemo(() => {
+    const _activeItems = itemsProp.filter((item) => activeKeys.includes(item.style))
+
+    let _items = itemsProp
+
+    if (_activeItems.length === 0 && _items.length > 1) {
+      _items = _items.concat([emptyStyle])
+      _activeItems.push(emptyStyle)
+    }
+
+    return {activeItems: _activeItems, items: _items}
+  }, [activeKeys, itemsProp])
+
+  const menuButtonText = useMemo(() => {
+    if (activeItems.length > 1) return 'Multiple'
+    if (activeItems.length === 1) return activeItems[0].title
+    return emptyStyle.title
+  }, [activeItems])
+
+  // Set focus back into the editor when the new value get's in
+  useEffect(() => {
+    if (changed) {
+      PortableTextEditor.focus(editor)
+      setChanged(false)
+    }
+  }, [activeItems, changed, editor])
+
+  const handleChange = useCallback(
+    (item: BlockStyleItem): void => {
+      if (focusBlock && item.style !== focusBlock.style) {
+        PortableTextEditor.toggleBlockStyle(editor, item.style)
+      }
+      setChanged(true)
+    },
+    [editor, focusBlock]
+  )
+
+  const renderOption = useCallback((style: string, title: string) => {
+    const hasTextStyle = TEXT_STYLE_KEYS.includes(style)
+
+    if (hasTextStyle) {
+      return TEXT_STYLE_OPTIONS[style](title)
+    }
```

Comment:
Refactor to prevent false positive in the future?
```suggestion
    const renderStyle = TEXT_STYLE_OPTIONS[style]

    if (renderStyle) {
      return renderStyle(title)
    }
```